gno_github_agent/clientql/client.go

```diff
@@ -0,0 +1,142 @@
+package clientql
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/TERITORI/gh-verify-agent/db"
+	"github.com/TERITORI/gh-verify-agent/gnoindexerql"
+	"github.com/TERITORI/gh-verify-agent/signer"
+	"go.uber.org/zap"
+	"gorm.io/gorm"
+)
+
+type IndexerQL struct {
+	gqlClient       graphql.Client
+	db              *gorm.DB
+	logger          *zap.SugaredLogger
+	signer          *signer.Signer
+	verifyRealmPath string
+}
+
+func New(graphqlEndpoint string, db *gorm.DB, logger *zap.SugaredLogger, gnoSigner *signer.Signer, verifyRealmPath string) *IndexerQL {
+	gqlClient := graphql.NewClient(graphqlEndpoint, nil)
+	return &IndexerQL{gqlClient: gqlClient, db: db, logger: logger, signer: gnoSigner, verifyRealmPath: verifyRealmPath}
+}
+
+func (client *IndexerQL) DealWithVerifications() error {
+	lastBlock, err := client.getLastTreatedBlock()
+	if err != nil {
+		return err
+	}
+
+	validationRequests, err := gnoindexerql.GetValidationRequests(context.Background(), client.gqlClient, lastBlock, client.verifyRealmPath)
+	if err != nil {
+		return err
+	}
+	client.logger.Infof("validation requests: %d\n", len(validationRequests.Transactions))
+
+	for _, validationRequest := range validationRequests.Transactions {
+		for _, responseEvent := range validationRequest.Response.Events {
+			switch event := responseEvent.(type) {
+			case *gnoindexerql.GetValidationRequestsTransactionsTransactionResponseEventsGnoEvent:
+				client.logger.Infof("args %v\n", event.Attrs)
+
+				err := client.dealWithVerification(event, validationRequest.Block_height)
+				if err != nil {
+					client.logger.Errorf("failed to deal with verification: %s", err.Error())
+					continue
+				}
+
+			default:
+				client.logger.Errorf("unexpected event type: %T", event)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (client *IndexerQL) getLastTreatedBlock() (int, error) {
+	var verification db.Verification
+	err := client.db.Model(&db.Verification{}).Where("status = ?", string(db.VerificationStatusVerified)).Order("id desc").First(&verification).Error
+
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return 0, nil
+		}
+
+		return 0, nil
+	}
+	return verification.BlockHeight, err
+}
+
+func (client *IndexerQL) dealWithVerification(event *gnoindexerql.GetValidationRequestsTransactionsTransactionResponseEventsGnoEvent, blockHeight int) error {
+	var handle string
+	var callerAddress string
+	for _, attr := range event.Attrs {
+		if attr.Key == "handle" {
+			handle = attr.Value
+		}
+		if attr.Key == "from" {
```

Comment:
we should use `else if` to prevent one more unused comparison

Reply:
fixed on f4a2a1f469fa31fc20f1534f07e68ae1d7408f89